csrc/optimization/layout_inference.cpp

```diff
@@ -0,0 +1,236 @@
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/all_nodes.h>
+#include <ir/utils.h>
+#include <optimization/layout_inference.h>
+
+namespace nvfuser {
+
+namespace {
+
+class MemoryFormatInferencer : public OptOutConstDispatch {
+ public:
+  MemoryFormatInferencer(
+      std::unordered_map<const TensorView*, MemoryFormat>& format_map)
+      : format_map_(format_map) {}
+
+ private:
+  void handle(const UnaryOp*) override;
+  void handle(const BinaryOp*) override;
+  void handle(const BroadcastOp*) override;
+  // TODO: Add more propagation rules
+  //   void handle(const Reduction*) override;
+  //   void handle(const LoadStoreOp*) override;
+  //   void handle(const SqueezeOp*) override;
+  //   void handle(const ExpandOp*) override;
+
+ private:
+  // format_map_ records the stride order (memory format) of each TensorView.
+  // Since it only handles permutation from a rfactor domain to allocation
+  // domain, it can be interpreted as:
+  //
+  // e.g. TV0 rfactor domain [i0, i1, i2]
+  //            alloc domain [i0, i2, i1]
+  //           memory format   0,  2,  1
+  std::unordered_map<const TensorView*, MemoryFormat>& format_map_;
+};
+
+// UnaryOp propagation forward memory format from input to output
+void MemoryFormatInferencer::handle(const UnaryOp* op) {
+  TensorView* out = dynamic_cast<TensorView*>(op->out());
+  if (out == nullptr) {
+    return;
+  }
+  TensorView* in = op->in()->as<TensorView>();
+  if (auto iter = format_map_.find(in); iter != format_map_.end()) {
+    format_map_[out] = iter->second;
+  }
+}
+
+// BinaryOp propagation tries to merge the memory format of both inputs
+//
+//   1. when there's only one operand has a recorded memory format, it forwards
+//   that.
+//   2. When both tensor have recorded memory format. It breaks tie based on the
+//   innermost dimension. whichever operand has a "better match" dominates the
+//   output format, where a "better match" meaning "less broadcast dimensions on
+//   the inner dimension"
+//
+// e.g.
+//   lhs TV0 rfactor_dom [i0, i1, b2]
+//                         0   2   1
+//   rhs TV0 rfactor_dom [i3, i4, b5]
+//                         0   1   2
+//   if we go from innermost to outermost order:
+//       TV0 has i1 -> b2 -> i0
+//       TV1 has b5 -> i4 -> i3
+//   we see that TV0 encounters a non-broadcast iter domain first, so TV0 is the
+//   dominating tensor. We'll produce an output with stride order identical to
+//   that of TV0 in the record.
+//   In the event of a tie, we'll just propagate the memory format of lhs.
+void MemoryFormatInferencer::handle(const BinaryOp* op) {
+  TensorView* out = dynamic_cast<TensorView*>(op->out());
+  if (out == nullptr) {
+    return;
+  }
+  TensorView* lhs = dynamic_cast<TensorView*>(op->lhs());
+  TensorView* rhs = dynamic_cast<TensorView*>(op->rhs());
+  if (lhs == nullptr) {
```

Comment:
OOC, when would this happen? `out` is a TensorView but `lhs` or `rhs` isn't?

Reply:
Something like `T1 = T0 + 3`?

Reply:
Sure. I thought 3 would be a rank-0 TensorView, but looks like it's going to be a Scalar.